lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -2760,7 +2760,7 @@ func (d *zfs) mountVolumeSnapshot(snapVol Volume, snapshotDataset string, mountP
 				}
 
 				// Delete on revert.
-				revert.Add(func() { _ = d.DeleteVolume(tmpVol, op) })
+				revert.Add(func() { _ = d.deleteDatasetRecursive(dataset) })
```

Comment:
Why is this needed?

Reply:
`tmpVol` is never actually created so running `d.DeleteVolume(tmpVol, op)` won't do anything. What we do is run `zfs clone <snapshotDataset> <dataset>` which we need to clean up.

Reply:
OK I think I'll hold off reviewing this further until this PR is split out a bit, as it feels like this PR is trying to address too many things at this time and its tricky to see how its all fitting together. Thanks.